antlr_ast/ast.py

```diff
@@ -283,3 +176,505 @@ def reportContextSensitivity(
+        return self.terminal_visitor(terminal)
+
+    def visit_str(self, string):
+        return self.terminal_visitor(string)
+
+    def terminal_visitor(self, terminal):
+        """Helper to transparently handle Terminal"""
+        return terminal
+
+
+def simplify_tree(tree, unpack_lists=True, in_list=False):
+    """Recursively unpack single-item lists and objects where fields and labels only reference a single child
+
+    :param tree: the tree to simplify (mutating!)
+    :param unpack_lists: whether single-item lists should be replaced by that item
+    :param in_list: this is used to prevent unpacking a node in a list as AST visit can't handle nested lists
+    """
+    # TODO: copy (or (de)serialize)? outside this function?
+    if isinstance(tree, BaseNode) and not isinstance(tree, Terminal):
+        used_fields = [field for field in tree._fields if getattr(tree, field, False)]
+        if len(used_fields) == 1:
+            result = getattr(tree, used_fields[0])
+        else:
+            result = None
+        if len(used_fields) != 1 or isinstance(tree, AliasNode) or (in_list and isinstance(result, list)):
+            result = tree
+            for field in tree._fields:
+                setattr(result, field, simplify_tree(getattr(tree, field), unpack_lists=unpack_lists))
+            return result
+        assert result is not None
+    elif isinstance(tree, list) and len(tree) == 1 and unpack_lists:
+        result = tree[0]
+    else:
+        if isinstance(tree, list):
+            result = [simplify_tree(el, unpack_lists=unpack_lists, in_list=True) for el in tree]
+        else:
+            result = tree
+        return result
+
+    return simplify_tree(result, unpack_lists=unpack_lists)
+
+
+class BaseAstVisitor(ParseTreeVisitor):
+    """Visitor that creates a high level tree
+
+    ~ ANTLR tree serializer
+    + automatic node creation using field and label detection
+    + alias nodes can work on tree without (ANTLR) visitor
+
+    Used from BaseAstVisitor: visitTerminal, visitErrorNode
+
+    TODO:
+     - [done] support labels
+     - [done] make compatible with AST: _fields = () (should only every child once)
+     - [done] include child_index to filter unique elements + order
+     - [done] memoize dynamic classes, to have list + make instance checks work
+     - [done] tree simplification as part of AliasNode
+     - [done] flatten nested list (see select with dynamic clause ordering)
+     - combine terminals / error nodes
+     - serialize highlight info
+     - make compatible with AstNode & AstModule in protowhat (+ shellwhat usage: bashlex + osh parser)
+         - combining fields & labels dicts needed?
+     - use exact ANTLR names in _rules (capitalize name without changing other casing)
+     - add labels to _fields if not overlapping with fields from rules
+     - [done] eliminate overhead of alias parsing (store ref to child index, get children on alias access)
```

Comment:
This is what I originally wrote down so I would think about it, but I didn't update it to explain what I did and why.

**The current implementation**

- visits children
- finds which child indices fields and labels refer to
- makes the correct children available under fields and labels (but indices are replaced by the Python object on initialisation, not at access time)

**Why?**

Since labels and fields can overlap, from the moment you have a node where that is the case, you would parse the overlapping subtree in two places (and every time such an overlap occurs, the number of parallel versions of the subtree at that point doubles.

When serialising, not using references for fields and labels would increase the size and becomes less clear for complex trees.
Besides more visiting work, it also requires more memory and a bit more work for gzip.

But the most important reason to parse children and then construct the fields and labels from the children at the correct indices, is that they are the same objects no matter how you access them (through a field or label) and equality checks will work as you would expect them to.